include/EASTL/sort.h

```diff
@@ -213,6 +213,54 @@ namespace eastl
 	/// inner for loop with a call to upper_bound, which would be faster.
 	///
 	template <typename BidirectionalIterator, typename StrictWeakOrdering>
+	void insertion_sort(
+	    BidirectionalIterator first,
+	    BidirectionalIterator last,
+	    StrictWeakOrdering compare,
+	    typename eastl::enable_if<
+	        eastl::is_move_constructible<typename eastl::iterator_traits<BidirectionalIterator>::value_type>::value>::
+	        type* = 0)
+	{
+		typedef typename eastl::iterator_traits<BidirectionalIterator>::value_type value_type;
+
+		if(first != last) // if the range is non-empty...
+		{
+			BidirectionalIterator iCurrent, iNext, iSorted = first;
+
+			for(++iSorted; iSorted != last; ++iSorted)
+			{
+				value_type temp( eastl::move(*iSorted) );
+
+				iNext = iCurrent = iSorted;
+
+				// Note: The following loop has a problem: it can decrement iCurrent to before 'first'.
+				// It doesn't dereference the iterator, but std STL disallows that operation. This isn't 
+				// a problem for EASTL containers and ranges, as they support a single decrement of first,
+				// but std STL iterators may have a problem with it. Dinkumware STL, for example, will assert.
+				// To do: Fix this loop to not decrement like so.
+				for(--iCurrent; (iNext != first) && compare(temp, *iCurrent); --iNext, --iCurrent)
+				{
+					EASTL_VALIDATE_COMPARE(!compare(*iCurrent, temp)); // Validate that the compare function is sane.
+					*iNext = eastl::move(*iCurrent);
+				}
+
+				*iNext = eastl::move(temp);
+			}
+		}
+	} // insertion_sort
+
+
+	/// insertion_sort
+	///
+	/// Since insertion_sort requires that the data be addressed with a BidirectionalIterator and 
+	/// not the more flexible RandomAccessIterator, we implement the sort by doing a for loop within
+	/// a for loop. If we were to specialize this for a RandomAccessIterator, we could replace the
+	/// inner for loop with a call to upper_bound, which would be faster.
+	///
+	template <typename BidirectionalIterator,
+	          typename StrictWeakOrdering,
+	          typename = typename eastl::enable_if<!eastl::is_move_constructible<
+	              typename eastl::iterator_traits<BidirectionalIterator>::value_type>::value>::type>
 	void insertion_sort(BidirectionalIterator first, BidirectionalIterator last, StrictWeakOrdering compare)
 	{
 		typedef typename eastl::iterator_traits<BidirectionalIterator>::value_type value_type;
```

Comment:
Let me try that out. That would certainly reduce code duplication. My reading of how it works with with respect to casting and rvalue to lvalue seems to indicate that this would be viable.